core/state_processor.go

```diff
@@ -713,6 +724,79 @@ func (p *StateProcessor) Process(block *types.WorkObject, batch ethdb.Batch) (ty
 				timeDelta := time.Since(startTimeEtx)
 				timeCoinbase += timeDelta
 				continue
+			} else if etx.EtxType() == types.ConversionRevertType {
+				// checking to make sure that the conversions that have to be
+				// refunded and processed before processing etxs from other
+				// zones
+				to := tx.To()
+				var sender common.Address
+				if to.IsInQuaiLedgerScope() {
+					sender = common.BytesToAddress(tx.Data()[2:22], p.hc.NodeLocation())
+				} else {
+					sender = tx.ETXSender()
+				}
+
+				// If the sender is in Quai ledger scope and to is in Qi,
+				// the reverted transaction should add the balance back to
+				// the original sender (Quai ledger)
+				if sender.IsInQuaiLedgerScope() && to.IsInQiLedgerScope() {
+					senderInternal, err := sender.InternalAddress()
+					if err != nil {
+						return nil, nil, nil, nil, 0, 0, 0, nil, nil, err
+					}
+					statedb.AddBalance(senderInternal, etx.Value())
+					*usedGas += params.QiToQuaiConversionGas
+					totalEtxGas += params.QiToQuaiConversionGas
+					receipt := &types.Receipt{Type: tx.Type(), Status: types.ReceiptStatusSuccessful, GasUsed: params.QiToQuaiConversionGas, TxHash: tx.Hash()}
+					receipts = append(receipts, receipt)
+					continue
+				}
+				if sender.IsInQiLedgerScope() && to.IsInQuaiLedgerScope() {
+					value := etx.Value()
+					// Lock the refund for ConversionLockPeriod
+					lock := new(big.Int).Set(new(big.Int).Add(block.Number(nodeCtx), big.NewInt(int64(params.ConversionLockPeriod))))
+					txGas := etx.Gas()
+					denominations := misc.FindMinDenominations(value)
+					outputIndex := uint16(0)
+					total := big.NewInt(0)
+					// Iterate over the denominations in descending order
+					for denomination := types.MaxDenomination; denomination > types.MaxTrimDenomination; denomination-- {
+
+						for j := uint64(0); j < denominations[uint8(denomination)]; j++ {
+							if txGas < params.CallValueTransferGas || outputIndex >= types.MaxOutputIndex {
+								break
+							}
+							if err := gp.SubGas(params.CallValueTransferGas); err != nil {
+								return nil, nil, nil, nil, 0, 0, 0, nil, nil, err
+							}
+							txGas -= params.CallValueTransferGas
+							*usedGas += params.CallValueTransferGas    // In the future we may want to determine what a fair gas cost is
+							totalEtxGas += params.CallValueTransferGas // In the future we may want to determine what a fair gas cost is
+							utxo := types.NewUtxoEntry(types.NewTxOut(uint8(denomination), sender.Bytes(), lock))
```

Comment:
You can actually ignore denomination 0 utxos entirely because they will get trimmed when they unlock

Reply:
yeah that is true, but no need to change that right then, feels like an optional thing, but can add if you insist

Reply:
actually its already taken care off, we only give until Qi out, any lower denomination is not added as an utxo

Reply:
check the for loop, it stops at the MaxTrimDenomination